lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      std::string s{parser::ToUpperCaseLetters(**charConst)};
+      if (specKind == SpecifierKind::Access) {
+        hasKnownAccess_ = true;
+        hasAccessDirect_ = s == "DIRECT"s;
+        hasAccessStream_ = s == "STREAM"s;
+      }
+      CheckStringValue(specKind, **charConst, parser::FindSourceLocation(spec));
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::Newunit &) {
+  SetSpecifier(SpecifierKind::Newunit);
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::Recl &spec) {
+  SetSpecifier(SpecifierKind::Recl);
+  if (const auto *expr{GetExpr(spec)}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    const std::optional<std::int64_t> &recl{evaluate::ToInt64(foldExpr)};
+    if (recl && *recl <= 0) {
+      context_.Say(parser::FindSourceLocation(spec),
+          "RECL value (%d) must be positive"_err_en_US,
+          std::move(*recl));  // 12.5.6.15
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::EndLabel &spec) {
+  SetSpecifier(SpecifierKind::End);
+}
+
+void IoChecker::Enter(const parser::EorLabel &spec) {
+  SetSpecifier(SpecifierKind::Eor);
+}
+
+void IoChecker::Enter(const parser::ErrLabel &spec) {
+  SetSpecifier(SpecifierKind::Err);
+}
+
+void IoChecker::Enter(const parser::FileUnitNumber &spec) {
+  SetSpecifier(SpecifierKind::Unit);
+  hasNumberUnit_ = true;
+}
+
+void IoChecker::Enter(const parser::Format &spec) {
+  SetSpecifier(SpecifierKind::Fmt);
+  if (std::get_if<parser::Star>(&spec.u)) {
+    hasStarFormat_ = true;
+  } else if (std::get_if<parser::Label>(&spec.u)) {
+    // Format statement format should be validated elsewhere.
+    hasLabelFormat_ = true;
+  } else {
+    hasCharFormat_ = true;
+    // TODO: validate compile-time constant format -- 12.6.2.2
+  }
+}
+
+void IoChecker::Enter(const parser::IdExpr &spec) {
+  SetSpecifier(SpecifierKind::Id);
+}
+
+void IoChecker::Enter(const parser::IdVariable &spec) {
+  SetSpecifier(SpecifierKind::Id);
+  auto expr{AnalyzeExpr(context_, spec.v)};
+  CHECK(expr);
+  int kind{expr->GetType()->kind};
+  int defaultKind{
+      context_.defaultKinds().GetDefaultKind(TypeCategory::Integer)};
+  if (kind < defaultKind) {
+    context_.Say(
+        "ID kind (%d) is smaller than default INTEGER kind (%d)"_err_en_US,
+        std::move(kind), std::move(defaultKind));  // C1229
+  }
+}
+
+void IoChecker::Enter(const parser::InputItem &spec) {
+  hasDataList_ = true;
+  if (const parser::Variable * var{std::get_if<parser::Variable>(&spec.u)}) {
+    const parser::Name &name{GetLastName(*var)};
+    if (auto *details{name.symbol->detailsIf<ObjectEntityDetails>()}) {
+      if (details->IsAssumedSize()) {
+        // This check may be superseded by C928 or C1002.
+        context_.Say(name.source,  // C1231
+            "'%s' is an assumed size array"_err_en_US, name.ToString().data());
```

Comment:
`grep` shows 77 instances of `data()` in `lib/` vs 30 instances of `c_str()`.  We should pick one and stick with it.  They're equivalent in C++17 but both can be misunderstood -- when I read `data()`, I worry about null termination, and when I read `c_str()`, I worry about whether a temp is being created and how long the pointer will be valid.  Neither is a real concern and so I'm not sure which is best.

Reply:
I agree we should pick one to use consistently. My preference would be `c_str()` but either is fine.

Even better would be a way to pass `Name` or `CharBlock` arguments to the `MessageFormattedText` constructor and have it do the conversion. That would allow us to eliminate a lot of instances of `.ToString().data()` and `.ToString().c_str()`.

Reply:
I've switched to c_str().